build/wpt_test.bzl

```diff
@@ -9,34 +9,88 @@
 
 load("//:build/wd_test.bzl", "wd_test")
 
-def wpt_test(name, wpt_directory, test_js):
-    test_gen_rule = "{}@_wpt_test_gen".format(name)
-    _wpt_test_gen(
-        name = test_gen_rule,
+def wpt_test(name, wpt_directory, test_config):
+    js_test_gen_rule = "{}@_wpt_js_test_gen".format(name)
+    _wpt_js_test_gen(
+        name = js_test_gen_rule,
         test_name = name,
         wpt_directory = wpt_directory,
-        test_js = test_js,
+        test_config = test_config,
+    )
+
+    wd_test_gen_rule = "{}@_wpt_wd_test_gen".format(name)
+    _wpt_wd_test_gen(
+        name = wd_test_gen_rule,
+        test_name = name,
+        wpt_directory = wpt_directory,
+        test_config = test_config,
+        test_js_generated = js_test_gen_rule,
     )
 
     wd_test(
         name = "{}".format(name),
-        src = test_gen_rule,
+        src = wd_test_gen_rule,
         args = ["--experimental"],
         data = [
             "//src/wpt:wpt-test-harness",
-            test_js,
+            test_config,
+            js_test_gen_rule,
             wpt_directory,
             "//src/workerd/io:trimmed-supported-compatibility-date.txt",
         ],
     )
 
-def _wpt_test_gen_impl(ctx):
+def _wpt_js_test_gen_impl(ctx):
+    src = ctx.actions.declare_file("{}-test.generated.js".format(ctx.attr.test_name))
+    ctx.actions.write(
+        output = src,
+        content = WPT_JS_TEST_TEMPLATE.format(
+            test_config = ctx.file.test_config.basename,
+            cases = generate_external_cases(ctx.attr.wpt_directory.files),
+        ),
+    )
+
+    return DefaultInfo(
+        files = depset([src]),
+    )
+
+def generate_external_cases(files):
+    result = []
+
+    for file in files.to_list():
+        if file.extension == "js":
+            entry = """export const {} = run(config, '{}');""".format(test_case_name(file.basename), file.basename)
+        else:
+            continue
+        result.append(entry)
+
+    return "\n".join(result)
+
+def test_case_name(filename):
+    words = (filename
+        .removesuffix(".js")
+        .removesuffix(".any")
+        .replace(".", "-")
+        .split("-"))
```

Comment:
```suggestion
        .split("."))
```

Reply:
If I literally applied this, I'd end up with invalid JS identifiers I think

Reply:
you are replacing . with `-` and later splitting with `-`. No need to traverse the string twice, where you can just split by `.` unless there is something I'm missing, which in that case i think it requires adding a comment.

Reply:
Ah! The original file names contain both `-` and `.` like `'url-setters-a-area.window.js`. I'm trying to turn this into a nice-ish JS identifier like `urlSettersAAreaWindow`. I'll add a comment to clarify.